File: grpc-xds/control-plane-java/src/main/java/com/google/examples/xds/controlplane/applications/Application.java
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.examples.xds.controlplane.applications;

import java.util.Collection;
import java.util.List;
import org.jetbrains.annotations.NotNull;

/**
 * Represents an application, e.g., a gRPC server, that clients discover using xDS.
 *
 * @param namespace the Kubernetes Namespace where the application is running
 * @param serviceAccountName the name of the Kubernetes Service Account used by the application
 * @param name the name that clients should use to connect to the application, following the <code>
 *     xds:///</code> scheme prefix. For instance, if <code>name</code> is <code>
 *     greeter-leaf</code>, clients connect using the {@link io.grpc.NameResolver}-compliant URI
 *     <code>xds:///greeter-leaf</code>
 * @param pathPrefix URL path prefix for the service exposed by the application. For a single route,
 *     the path prefix can be an empty string. In general, for a gRPC service, the path prefix will
 *     be in the format <code>/[package].[service]/</code>, e.g., <code>
 *     /helloworld.Greeter/</code>
 * @param servingPort serving port of the server
 * @param healthCheckPort health check port of the server, can be the same as the serving port
 * @param endpoints application server endpoints
 */
public record Application(
    @NotNull String namespace,
    @NotNull String serviceAccountName,
    @NotNull String name,
    @NotNull String pathPrefix,
    int servingPort,
    @NotNull String servingProtocol,
    int healthCheckPort,
    @NotNull String healthCheckProtocol,
    @NotNull Collection<ApplicationEndpoint> endpoints) {

  /**
   * Convenience constructor for use in the common scenario where the Kubernetes ServiceAccount and
   * application share the same name.
   *
   * @param namespace the Kubernetes Namespace
   * @param name the Kubernetes ServiceAccount name and application name - which must match to use
   *     this constructor!
   * @param servingPort serving port of the server
   * @param healthCheckPort health check port of the server, can be the same as the serving port
   * @param endpoints application server endpoints
   */
  public Application(
      @NotNull String namespace,
      @NotNull String name,
      int servingPort,
      @NotNull String servingProtocol,
      int healthCheckPort,
      @NotNull String healthCheckProtocol,
      @NotNull Collection<ApplicationEndpoint> endpoints) {
    this(
        namespace,
        name,
        name,
        "",
        servingPort,
        servingProtocol,
        healthCheckPort,
        healthCheckProtocol,
        List.copyOf(endpoints));
  }
}

File: grpc-xds/control-plane-go/pkg/applications/application.go
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package applications

import (
	"slices"
	"strings"
)

// Application represents an application, e.g., a gRPC server, that clients discover using xDS.
type Application struct {
	Namespace           string
	ServiceAccountName  string
	Name                string
	PathPrefix          string
	ServingPort         uint32
	ServingProtocol     string
	HealthCheckPort     uint32
	HealthCheckProtocol string
	Endpoints           []ApplicationEndpoints
}

// NewApplication is a convenience function that creates a Application where the
// k8s ServiceAccount and the application share the same name.
func NewApplication(namespace string, name string, servingPort uint32, servingProtocol string, healthCheckPort uint32, healthCheckProtocol string, endpoints []ApplicationEndpoints) Application {
	endpointsCopy := make([]ApplicationEndpoints, len(endpoints))
	copy(endpointsCopy, endpoints)
	slices.SortFunc(endpointsCopy, func(a ApplicationEndpoints, b ApplicationEndpoints) int {
		return a.Compare(b)
	})
	return Application{
		Namespace:           namespace,
		ServiceAccountName:  name,
		Name:                name,
		PathPrefix:          "",
		ServingPort:         servingPort,
		ServingProtocol:     servingProtocol,
		HealthCheckPort:     healthCheckPort,
		HealthCheckProtocol: healthCheckProtocol,
		Endpoints:           endpointsCopy,
	}
}

// Compare assumes that the list of endpoints is sorted,
// as done in `NewApplication()`.
func (a Application) Compare(b Application) int {
	if a.Namespace != b.Namespace {
		return strings.Compare(a.Namespace, b.Namespace)
	}
	if a.ServiceAccountName != b.ServiceAccountName {
		return strings.Compare(a.ServiceAccountName, b.ServiceAccountName)
	}
	if a.Name != b.Name {
		return strings.Compare(a.Name, b.Name)
	}
	if a.PathPrefix != b.PathPrefix {
		return strings.Compare(a.PathPrefix, b.PathPrefix)
	}
	if a.ServingPort != b.ServingPort {
		return int(a.ServingPort - b.ServingPort)
	}
	if a.ServingProtocol != b.ServingProtocol {
		return strings.Compare(a.ServingProtocol, b.ServingProtocol)
	}
	if a.HealthCheckPort != b.HealthCheckPort {
		return int(a.HealthCheckPort - b.HealthCheckPort)
	}
	if a.HealthCheckProtocol != b.HealthCheckProtocol {
		return strings.Compare(a.HealthCheckProtocol, b.HealthCheckProtocol)
	}
	return slices.CompareFunc(a.Endpoints, b.Endpoints,
		func(e ApplicationEndpoints, f ApplicationEndpoints) int {
			return e.Compare(f)
		})
}

// Equal assumes that the list of endpoints is sorted,
// as done in `NewApplication()`.
func (a Application) Equal(b Application) bool {
	return a.Compare(b) == 0
}
